crates/prune/src/pruner.rs

```diff
@@ -146,69 +180,76 @@ impl<DB: Database> Pruner<DB> {
-                    .duration_seconds
-                    .record(segment_start.elapsed());
-
-                self.metrics
-                    .get_prune_segment_metrics(segment.segment())
-                    .highest_pruned_block
-                    .set(to_block as f64);
-
-                done = done && output.done;
-                delete_limit = delete_limit.saturating_sub(output.pruned);
-
-                debug!(
-                    target: "pruner",
-                    segment = ?segment.segment(),
-                    ?purpose,
-                    %to_block,
-                    ?prune_mode,
-                    %output.pruned,
-                    "Segment pruning finished"
-                );
-
-                if output.pruned > 0 {
-                    stats.insert(
-                        segment.segment(),
-                        (PruneProgress::from_done(output.done), output.pruned),
+                    let segment_start = Instant::now();
+                    let previous_checkpoint = provider.get_prune_checkpoint(segment.segment())?;
+                    let output = segment
+                        .prune(provider, PruneInput { previous_checkpoint, to_block, limiter })?;
+                    if let Some(checkpoint) = output.checkpoint {
+                        segment.save_checkpoint(
+                            provider,
+                            checkpoint.as_prune_checkpoint(prune_mode),
+                        )?;
+                    }
+                    self.metrics
+                        .get_prune_segment_metrics(segment.segment())
+                        .duration_seconds
+                        .record(segment_start.elapsed());
+
+                    self.metrics
+                        .get_prune_segment_metrics(segment.segment())
+                        .highest_pruned_block
+                        .set(to_block as f64);
+
+                    done = done && output.progress.is_done();
+                    limiter.increment_deleted_units_count_by(output.pruned);
+
+                    debug!(
+                        target: "pruner",
+                        segment = ?segment.segment(),
+                        ?purpose,
+                        %to_block,
+                        ?prune_mode,
+                        %output.pruned,
+                        "Segment pruning finished"
                     );
+
+                    if output.pruned > 0 {
+                        // sets `is_timed_out`
+                        let _ = limiter.at_limit();
+
+                        stats.insert(segment.segment(), (output.progress, output.pruned));
+                    }
+                } else {
+                    debug!(target: "pruner", segment = ?segment.segment(), ?purpose, "Nothing to prune for the segment");
+                }
+
+                if limiter.at_limit() {
+                    done = false;
+                    break
```

Comment:
unsure if that's correct, what if we actually finished (`done = true`), but also hit the limit?